### PR TITLE
[WIP] Added function to eulerize simple graphs to algorithms.euler

### PR DESCRIPTION
**Pull request by mtrenfield**
As discussed in [issue 2869](https://github.com/networkx/networkx/issues/2869), I have implemented a function to transform a graph into an Eulerian graph along the lines mentioned in [these lecture notes](http://www.math.unl.edu/%7Ebharbourne1/M203Spr07/EulerizationHandout.pdf) and using the method described [here](https://cs.stackexchange.com/questions/9126/is-there-a-non-brute-force-algorithm-for-eulerization-of-graphs) to find an efficient set of paths.

I have also written test cases for Eulerization of Erdos-Renyi graphs and some complete graphs.

I am labeling this as a WIP because this would be my first commit to NetworkX and expect that there may be things I haven't thought of or things I have done wrong.  I have also not implemented this for directed graphs.  This is because I use `nx.max_weight_matching` to find an Eulerization that doesn't require duplicated too many edges.  I am sure directed graphs can be Eulerized, but I can't see how to do this efficiently.

### CODE REVIEW

File: networkx/algorithms/euler.py [hunk 1]
```diff
@@ -181,3 +183,78 @@ def eulerian_circuit(G, source=None, keys=False):
...
+            if n != m:
+                Gp.add_edge(m,n,weight=1/len(P), path=P)
+
+    # find the minimum weight matching of edges in the weighted graph
+    best_matching = nx.Graph(list(nx.max_weight_matching(Gp)))
+
+    # duplicate each edge along each path in the set of paths in Gp
+    for m,n in best_matching.edges():
+        path = Gp[m][n]["path"]
+        G.add_edges_from([(path[i], path[i+1]) for i,_ in enumerate(path[:-1])])
```
Comment (dschult) [inline networkx/algorithms/euler.py]: There is a ```networkx.utils.pairwise``` function for this:

   G.add_edges_from(nx.utils.pairwise(path))

File: networkx/algorithms/tests/test_euler.py [hunk 2]
```diff
@@ -112,3 +112,15 @@ def test_multigraph_with_keys(self):
     @raises(nx.NetworkXError)
     def test_not_eulerian(self):
         f = list(eulerian_circuit(nx.complete_graph(4)))
+
+class TestEulerize(TestCase):
+
```
Comment (dschult) [inline networkx/algorithms/tests/test_euler.py]: You should probably add the corner cases like ```G=nx.Graph()``` and ```G=nx.empty_graph(3)```.
Also, test that it raises an error for a directed graph.

**Comment by dschult** [comment]
This look good!  
I made some comments connected to individual lines of code.

You should also bring the code style up to the PEP8 standards. Its easy as ```pip install pep8``` and then ```pep8 euler.py```.  Most of the trouble is typically spacing issues and 79 char limit on length of line.  You can also read the [PEP8 style guidelines in their official form](https://www.python.org/dev/peps/pep-0008/?). 

Also, the single-use function ```_simple_eulerize_simplegraph``` can be eliminated for simplicity by removing lines 230-233

The Codecov check (click on "Diff" heading of the details to see how Codecov thinks your tests are covering the code) is basically asking you to test the two NetworkXError cases.  It also asks you to test a case where ```n==m``` (which I think can't happen since you used ```combinations``` to create the data.

You should probably add simple tests that input a MultiGraph with multiedges to check that it is handled the way you want.  Just a simple one like a triangle with an extra edge would do it I think -- but maybe there's more there to test.

**Comment by mtrenfield** [comment]
How do you think the case of the null graph be handled?  It could be considered as trivially Eulerian and I could return nx.MultiGraph(), but I could also raise an exception.  I don't have a strong preference for one over the other, but I see that nx.is_connected(nx.Graph()) raises a NetworkXPointlessConcept exceptions, so perhaps I should, too, for the sake of consistency.  Thoughts?

**Comment by lexelby** [comment]
Amazing!  This is going to be super useful for [Ink/Stitch](https://github.com/inkstitch/inkstitch).  Thanks for your work!

Do you folks have any idea when this might make it into a stable release?

**Pull request merged**